verification/src/block_verifier.rs

```diff
@@ -521,3 +532,76 @@ impl<CP: ChainProvider + Clone> CommitVerifier<CP> {
         Ok(())
     }
 }
+
+#[derive(Clone)]
+pub struct BlockSizeVerifier<CP> {
+    provider: CP,
+}
+
+impl<CP: ChainProvider + Clone> BlockSizeVerifier<CP> {
+    pub fn new(provider: CP) -> Self {
+        BlockSizeVerifier { provider }
+    }
+
+    pub fn verify(&self, block: &Block) -> Result<(), Error> {
+        let fbb = &mut FlatBufferBuilder::new();
+        let block_offset = FbsBlock::build(fbb, block);
+        fbb.finish(block_offset, None);
+        if fbb.finished_data().len() as u64 > self.provider.consensus().max_block_bytes {
+            return Err(Error::ExceededMaximumBlockSize);
+        }
+        Ok(())
+    }
+}
+
+#[derive(Clone)]
+pub struct IncrementOccupiedCapacityVerifier<CP> {
+    provider: CP,
+}
+
+impl<CP: ChainProvider + CellProvider + Clone> IncrementOccupiedCapacityVerifier<CP> {
+    pub fn new(provider: CP) -> Self {
+        IncrementOccupiedCapacityVerifier { provider }
+    }
+
+    pub fn verify(&self, block: &Block) -> Result<(), Error> {
+        if let Some((inputs_occupied, outputs_occupied)) = self.calculate_occupied_capacity(block) {
+            if outputs_occupied.saturating_sub(inputs_occupied)
+                > self.provider.consensus().max_increment_occupied_capacity()
+            {
+                return Err(Error::ExceededMaximumIncrementOccupiedCapacity);
+            }
+        }
+        Ok(())
+    }
+
+    fn calculate_occupied_capacity(&self, block: &Block) -> Option<(Capacity, Capacity)> {
+        block.commit_transactions().iter().try_fold(
+            (0, 0),
+            |(inputs_occupied, outputs_occupied): (Capacity, Capacity), tx| {
```

Comment:
We can use `inputs_occupied` as inner accumulator to avoid each iteration allocation and simplify code like this:
```rust
            |(inputs_occupied, outputs_occupied): (Capacity, Capacity), tx| {
                tx.inputs().iter().try_fold(inputs_occupied, |acc, input|
                    self.provider
                        .cell(&input.previous_output)
                        .get_live()
                        .map(|output| acc.saturating_add(output.occupied_capacity() as Capacity))
                ).map(|inputs_occupied|
                    (
                        inputs_occupied,
                        outputs_occupied.saturating_add(
                            tx.outputs()
                                .iter()
                                .map(|output| output.occupied_capacity() as Capacity)
                                .sum(),
                        ),
                    )
                )
            },
```

Reply:
Ok, remove one unnecessary iterator, but I leave the `if` statement to make the code easier to understand.